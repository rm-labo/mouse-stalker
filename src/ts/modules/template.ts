import { MsOptions } from '../types'

export const getTemplate = (opt: MsOptions) => {
  return `
    <div class="${opt.mouseElementClassName}">
      <div class="${opt.pointerElementClassName}"></div>
      <div class="${opt.iconTargetElementClassName}"></div>
    </div>
    <style>
      :host {}
      [${opt.targetSelectorsDataName}]:hover {
        cursor: none;
      }
      .${opt.mouseElementClassName} {
        --st-transition-timing-function: cubic-bezier(.2,.8,.4,.9);

        pointer-events: none;
        position: fixed;
        will-change: transform;
        z-index: ${opt.zIndex};
        top: 0;
        left: 0;
        transform: translate(0,0);
        transition: transform 0.3s;
        transition-timing-function: var(--st-transition-timing-function);
        mix-blend-mode: difference;
        filter: invert(100%);
      }
      .${opt.pointerElementClassName} {
        position: absolute;
        top: 0;
        left: 0;
        z-index: 9999;
        border-radius: 50%;
        background-color: ${opt.pointerColor};
        width: ${opt.pointerSize}px;
        height: ${opt.pointerSize}px;
        margin-top: ${-1 * opt.pointerSize * 0.5}px;
        margin-left: ${-1 * opt.pointerSize * 0.5}px;
        transform: rotate(0deg) scale(1, 1);
        transition: transform 0s;
        transition: background 0.6s;
        transition-timing-function: var(--st-transition-timing-function);
      }
      .${opt.mouseElementClassName}::before {
        content: "";
        position: absolute;
        top: 50%;
        left: 50%;
        width: ${opt.pointerSize * opt.hoveringExpansionRatio}px;
        height: ${opt.pointerSize * opt.hoveringExpansionRatio}px;
        margin-top: ${-1 * opt.pointerSize * opt.hoveringExpansionRatio * 0.5}px;
        margin-left: ${-1 * opt.pointerSize * opt.hoveringExpansionRatio * 0.5}px;
        border-radius: 50%;
        border-width: ${opt.hoveringBorderWidth}px;
        border-style: solid;
        border-color: ${opt.pointerColor};
        transition: transform 0.9s;
        transition-timing-function: var(--st-transition-timing-function);
        transform: scale(0);
        transform-origin: center center;
      }
      .${opt.iconTargetElementClassName} {
        color: ${opt.pageBaseColor};
        font-size: ${opt.pointerSize * opt.hoveringExpansionRatio * 0.5}px;
        position: absolute;
        top: 0;
        left: 0;
        width: ${opt.pointerSize}px;
        height: ${opt.pointerSize}px;
        margin-top: ${-1 * (opt.pointerSize * 0.5)}px;
        margin-left: ${-1 * (opt.pointerSize * 0.5)}px;
        line-height: 1;
        transition: transform 0.9s;
        transition-timing-function: var(--st-transition-timing-function);
        transform: scale(0.8);
        animation-duration: 0.6s;
        opacity: 0;
        transform: translate(0, 0);
        display: flex;
        justify-content: center;
        align-items: center;
      }
      .${opt.iconTargetElementClassName} svg {
        position: absolute;
      }

      .${opt.mouseElementClassName}.has-icon {
        mix-blend-mode: normal;
        filter: invert(0);
      }
      .${opt.mouseElementClassName}.has-icon::before {
        background-color: ${opt.pointerColor};
      }

      @media (hover: hover) {
        .is-hover .${opt.pointerElementClassName}{
          transition: background 0.3s;
          background-color: transparent;
        }
        .${opt.mouseElementClassName}.is-hover::before {
          transition: transform 0.3s;
          transform: scale(1);
        }
        .${opt.mouseElementClassName}.is-hover .${opt.iconTargetElementClassName} {
          transform: scale(1);
          opacity: 1;
          animation: show 0.3s var(--st-transition-timing-function);
        }
        .${opt.mouseElementClassName}[${opt.hoveringIconAnimationName}-target="toTop"].is-hover .${
    opt.iconTargetElementClassName
  } {
          animation: toTop 0.3s var(--st-transition-timing-function);
        }
        .${opt.mouseElementClassName}[${opt.hoveringIconAnimationName}-target="toBottom"].is-hover .${
    opt.iconTargetElementClassName
  } {
          animation: toBottom 0.3s var(--st-transition-timing-function);
        }
        .${opt.mouseElementClassName}[${opt.hoveringIconAnimationName}-target="toRight"].is-hover .${
    opt.iconTargetElementClassName
  } {
          animation: toRight 0.3s var(--st-transition-timing-function);
        }
        .${opt.mouseElementClassName}[${opt.hoveringIconAnimationName}-target="toLeft"].is-hover .${
    opt.iconTargetElementClassName
  } {
          animation: toLeft 0.3s var(--st-transition-timing-function);
        }
      }
      
      @keyframes show {
        0% {
          opacity: 0;
          transform: scale(2);
        }
        100% {
          opacity: 1;
          transform: scale(1);
        }
      }
      @keyframes toBottom {
        0% {
          opacity: 0;
          transform: translate(0, -100%);
        }
        100% {
          opacity: 1;
          transform: translate(0, 0);
        }
      }
      @keyframes toTop {
        0% {
          opacity: 0;
          transform: translate(0, 100%);
        }
        100% {
          opacity: 1;
          transform: translate(0, 0);
        }
      }
      @keyframes toRight {
        0% {
          opacity: 0;
          transform: translate(-100%, 0);
        }
        100% {
          opacity: 1;
          transform: translate(0, 0);
        }
      }
      @keyframes toLeft {
        0% {
          opacity: 0;
          transform: translate(100%, 0);
        }
        100% {
          opacity: 1;
          transform: translate(0, 0);
        }
      }
    </style>
  `
}
