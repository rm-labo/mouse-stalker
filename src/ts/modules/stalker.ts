import { MsOptions } from '../types'
import { getTemplate } from './template'
import { renderHTML, loadIcons, iconExists } from './icons'

const animationNames = ['default', 'toTop', 'toRight', 'toBottom', 'toLeft'] as const

export const defaultOptions = {
  mouseElementClassName: 'mouse-stalker-root',
  pointerElementClassName: 'mouse-stalker-pointer',
  iconTargetElementClassName: 'mouse-stalker-icon-target',
  pageBaseColor: '#ffffff',
  pointerColor: 'rgba(0,0,0,1)',
  pointerSize: 24,
  zIndex: '9999',
  hoveringExpansionRatio: 2,
  hoveringBorderWidth: 1,
  scoped: false,
  targetSelectorsDataName: 'data-st-target',
  hoveringIconName: 'data-st-icon-name',
  hoveringIconAnimationName: 'data-st-icon-animation-name',
}

/**
 * Stalker
 */
export default class Stalker {
  private _opt: MsOptions
  private _rootElement: HTMLElement
  private _rootTarget: HTMLElement | ShadowRoot
  private _mouseElement: HTMLElement
  private _pointerElement: HTMLElement
  private _iconTargetElement: HTMLElement
  private _selectors: NodeListOf<Element>
  private _handleMousemove: (e: MouseEvent) => void
  private _handleMouseover: (e: Event) => void
  private _handleMouseout: (e: Event) => void

  static defaultOptions: MsOptions = defaultOptions

  constructor(private options: MsOptions) {
    this._opt = Object.assign(Stalker.defaultOptions, options)
  }

  init(): void {
    const iconAnimationDataName = `${this._opt.hoveringIconAnimationName}-target`
    const initIcons = (): void => {
      const _names: string[] = []
      const _iconTargets = this._selectors
      _iconTargets.forEach((e) => _names.push(e.getAttribute(this._opt.hoveringIconName)))
      _names.filter((e) => e)
      loadIcons(_names)
    }

    const initHoverTarget = (): void => {
      this._selectors.forEach((element) => {
        element.addEventListener('mouseenter', this._handleMouseover)
        element.addEventListener('mouseleave', this._handleMouseout)
      })
    }

    const showIcon = (event: Event): void => {
      const _target = event.target as HTMLElement
      const iconName = _target.getAttribute(this._opt.hoveringIconName)
      const animationName = _target.getAttribute(this._opt.hoveringIconAnimationName) as typeof animationNames[number]

      if (iconName) {
        if (!iconExists(iconName)) {
          loadIcons([iconName])
        }
        this._iconTargetElement.innerHTML = renderHTML(iconName)
        if (animationName) {
          if (!animationNames.includes(animationName)) {
            throw new Error(
              `Acceptable values for ${this._opt.hoveringIconAnimationName} are ${animationNames.toString()}.`
            )
          }
          this._mouseElement.setAttribute(iconAnimationDataName, animationName)
        }
        this._mouseElement.classList.add('has-icon')
      }
    }

    const hideIcon = (): void => {
      this._mouseElement.classList.remove('has-icon')
      this._iconTargetElement.innerHTML = null
      this._mouseElement.removeAttribute(iconAnimationDataName)
    }

    const renderElement = (): void => {
      this._rootElement = document.createElement('div')

      if (this._opt.scoped) {
        this._rootElement.attachShadow({ mode: 'open' })
        this._rootTarget = this._rootElement.shadowRoot
      } else {
        this._rootTarget = this._rootElement
      }

      this._rootTarget.innerHTML = getTemplate(this._opt)
      document.body.appendChild(this._rootElement)
      document.body.style.backgroundColor = this._opt.pageBaseColor

      this._mouseElement = this._rootTarget.querySelector(`.${this._opt.mouseElementClassName}`)
      this._pointerElement = this._rootTarget.querySelector(`.${this._opt.pointerElementClassName}`)
      this._iconTargetElement = this._rootTarget.querySelector(`.${this._opt.iconTargetElementClassName}`)
    }

    const mouseMove = (e: MouseEvent): void => {
      let _movementTimer: NodeJS.Timer | null = null
      const _mouse = {
        x: e.clientX,
        y: e.clientY,
        mx: e.movementX,
        my: e.movementY,
      }
      const _degrees = Math.atan2(_mouse.my, _mouse.mx) * (180 / Math.PI) + 90
      const _distance = Math.sqrt(Math.pow(_mouse.mx, 2) + Math.pow(_mouse.my, 2))
      const _scaleX = 1 - (_distance / 100) * 0.5
      const _scaleY = 1 + (_distance / 100) * 0.5

      this._mouseElement.style.transform = `translate3d(${_mouse.x}px, ${_mouse.y}px, 0)`
      this._pointerElement.style.transform = `rotate(${_degrees}deg) scale(${_scaleX}, ${_scaleY})`

      // Reset distortion when going off-screen
      clearTimeout(_movementTimer)
      _movementTimer = setTimeout(() => {
        if (this._pointerElement) {
          this._pointerElement.style.transform = 'rotate(0deg) scale(1, 1)'
        }
      }, 100)
    }

    this._selectors = document.querySelectorAll(`[${this._opt.targetSelectorsDataName}]`)

    this._handleMouseover = (event) => {
      showIcon(event)
      this._mouseElement.classList.add('is-hover')
    }

    this._handleMouseout = () => {
      hideIcon()
      this._mouseElement.classList.remove('is-hover')
    }

    this._handleMousemove = (event) => {
      mouseMove(event)
    }

    initIcons()
    initHoverTarget()
    renderElement()

    document.addEventListener('mousemove', this._handleMousemove)
  }

  destroy(): void {
    document.removeEventListener('mousemove', this._handleMousemove)
    this._selectors.forEach((element) => {
      element.removeEventListener('mouseenter', this._handleMouseover)
      element.removeEventListener('mouseleave', this._handleMouseout)
    })
    this._rootElement.remove()
    this._rootElement = null
    this._rootTarget = null
    this._mouseElement = null
    this._pointerElement = null
    this._iconTargetElement = null
    this._selectors = null
    this._handleMousemove = null
    this._handleMouseover = null
    this._handleMouseout = null
  }
}
