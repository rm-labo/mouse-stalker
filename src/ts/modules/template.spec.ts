import { getTemplate } from './template'
import { defaultOptions } from './stalker'

describe('getTemplate()', (): void => {
  it('toMatchSnapshot', () => {
    expect(getTemplate(defaultOptions)).toMatchSnapshot()
  })
})
