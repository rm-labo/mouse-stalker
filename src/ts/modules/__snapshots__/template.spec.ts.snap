// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getTemplate() getTemplate()のスナップショット 1`] = `
"
    <div class=\\"mouse-stalker-root\\">
      <div class=\\"mouse-stalker-pointer\\"></div>
      <div class=\\"mouse-stalker-icon-target\\"></div>
    </div>
    <style>
      :host {}
      [data-st-target]:hover {
        cursor: none;
      }
      .mouse-stalker-root {
        --st-transition-timing-function: cubic-bezier(.2,.8,.4,.9);

        pointer-events: none;
        position: fixed;
        will-change: transform;
        z-index: 9999;
        top: 0;
        left: 0;
        transform: translate(0,0);
        transition: transform 0.3s;
        transition-timing-function: var(--st-transition-timing-function);
        mix-blend-mode: difference;
        filter: invert(100%);
      }
      .mouse-stalker-pointer {
        position: absolute;
        top: 0;
        left: 0;
        z-index: 9999;
        border-radius: 50%;
        background-color: rgba(0,0,0,1);
        width: 24px;
        height: 24px;
        margin-top: -12px;
        margin-left: -12px;
        transform: rotate(0deg) scale(1, 1);
        transition: transform 0s;
        transition: background 0.6s;
        transition-timing-function: var(--st-transition-timing-function);
      }
      .mouse-stalker-root::before {
        content: \\"\\";
        position: absolute;
        top: 50%;
        left: 50%;
        width: 48px;
        height: 48px;
        margin-top: -24px;
        margin-left: -24px;
        border-radius: 50%;
        border-width: 1px;
        border-style: solid;
        border-color: rgba(0,0,0,1);
        transition: transform 0.9s;
        transition-timing-function: var(--st-transition-timing-function);
        transform: scale(0);
        transform-origin: center center;
      }
      .mouse-stalker-icon-target {
        color: #ffffff;
        font-size: 24px;
        position: absolute;
        top: 0;
        left: 0;
        width: 24px;
        height: 24px;
        margin-top: -12px;
        margin-left: -12px;
        line-height: 1;
        transition: transform 0.9s;
        transition-timing-function: var(--st-transition-timing-function);
        transform: scale(0.8);
        animation-duration: 0.6s;
        opacity: 0;
        transform: translate(0, 0);
        display: flex;
        justify-content: center;
        align-items: center;
      }
      .mouse-stalker-icon-target svg {
        position: absolute;
      }

      .mouse-stalker-root.has-icon {
        mix-blend-mode: normal;
        filter: invert(0);
      }
      .mouse-stalker-root.has-icon::before {
        background-color: rgba(0,0,0,1);
      }

      @media (hover: hover) {
        .is-hover .mouse-stalker-pointer{
          transition: background 0.3s;
          background-color: transparent;
        }
        .mouse-stalker-root.is-hover::before {
          transition: transform 0.3s;
          transform: scale(1);
        }
        .mouse-stalker-root.is-hover .mouse-stalker-icon-target {
          transform: scale(1);
          opacity: 1;
          animation: show 0.3s var(--st-transition-timing-function);
        }
        .mouse-stalker-root[data-st-icon-animation-name-target=\\"toTop\\"].is-hover .mouse-stalker-icon-target {
          animation: toTop 0.3s var(--st-transition-timing-function);
        }
        .mouse-stalker-root[data-st-icon-animation-name-target=\\"toBottom\\"].is-hover .mouse-stalker-icon-target {
          animation: toBottom 0.3s var(--st-transition-timing-function);
        }
        .mouse-stalker-root[data-st-icon-animation-name-target=\\"toRight\\"].is-hover .mouse-stalker-icon-target {
          animation: toRight 0.3s var(--st-transition-timing-function);
        }
        .mouse-stalker-root[data-st-icon-animation-name-target=\\"toLeft\\"].is-hover .mouse-stalker-icon-target {
          animation: toLeft 0.3s var(--st-transition-timing-function);
        }
      }
      
      @keyframes show {
        0% {
          opacity: 0;
          transform: scale(2);
        }
        100% {
          opacity: 1;
          transform: scale(1);
        }
      }
      @keyframes toBottom {
        0% {
          opacity: 0;
          transform: translate(0, -100%);
        }
        100% {
          opacity: 1;
          transform: translate(0, 0);
        }
      }
      @keyframes toTop {
        0% {
          opacity: 0;
          transform: translate(0, 100%);
        }
        100% {
          opacity: 1;
          transform: translate(0, 0);
        }
      }
      @keyframes toRight {
        0% {
          opacity: 0;
          transform: translate(-100%, 0);
        }
        100% {
          opacity: 1;
          transform: translate(0, 0);
        }
      }
      @keyframes toLeft {
        0% {
          opacity: 0;
          transform: translate(100%, 0);
        }
        100% {
          opacity: 1;
          transform: translate(0, 0);
        }
      }
    </style>
  "
`;

exports[`getTemplate() toMatchSnapshot 1`] = `
"
    <div class=\\"mouse-stalker-root\\">
      <div class=\\"mouse-stalker-pointer\\"></div>
      <div class=\\"mouse-stalker-icon-target\\"></div>
    </div>
    <style>
      :host {}
      [data-st-target]:hover {
        cursor: none;
      }
      .mouse-stalker-root {
        --st-transition-timing-function: cubic-bezier(.2,.8,.4,.9);

        pointer-events: none;
        position: fixed;
        will-change: transform;
        z-index: 9999;
        top: 0;
        left: 0;
        transform: translate(0,0);
        transition: transform 0.3s;
        transition-timing-function: var(--st-transition-timing-function);
        mix-blend-mode: difference;
        filter: invert(100%);
      }
      .mouse-stalker-pointer {
        position: absolute;
        top: 0;
        left: 0;
        z-index: 9999;
        border-radius: 50%;
        background-color: rgba(0,0,0,1);
        width: 24px;
        height: 24px;
        margin-top: -12px;
        margin-left: -12px;
        transform: rotate(0deg) scale(1, 1);
        transition: transform 0s;
        transition: background 0.6s;
        transition-timing-function: var(--st-transition-timing-function);
      }
      .mouse-stalker-root::before {
        content: \\"\\";
        position: absolute;
        top: 50%;
        left: 50%;
        width: 48px;
        height: 48px;
        margin-top: -24px;
        margin-left: -24px;
        border-radius: 50%;
        border-width: 1px;
        border-style: solid;
        border-color: rgba(0,0,0,1);
        transition: transform 0.9s;
        transition-timing-function: var(--st-transition-timing-function);
        transform: scale(0);
        transform-origin: center center;
      }
      .mouse-stalker-icon-target {
        color: #ffffff;
        font-size: 24px;
        position: absolute;
        top: 0;
        left: 0;
        width: 24px;
        height: 24px;
        margin-top: -12px;
        margin-left: -12px;
        line-height: 1;
        transition: transform 0.9s;
        transition-timing-function: var(--st-transition-timing-function);
        transform: scale(0.8);
        animation-duration: 0.6s;
        opacity: 0;
        transform: translate(0, 0);
        display: flex;
        justify-content: center;
        align-items: center;
      }
      .mouse-stalker-icon-target svg {
        position: absolute;
      }

      .mouse-stalker-root.has-icon {
        mix-blend-mode: normal;
        filter: invert(0);
      }
      .mouse-stalker-root.has-icon::before {
        background-color: rgba(0,0,0,1);
      }

      @media (hover: hover) {
        .is-hover .mouse-stalker-pointer{
          transition: background 0.3s;
          background-color: transparent;
        }
        .mouse-stalker-root.is-hover::before {
          transition: transform 0.3s;
          transform: scale(1);
        }
        .mouse-stalker-root.is-hover .mouse-stalker-icon-target {
          transform: scale(1);
          opacity: 1;
          animation: show 0.3s var(--st-transition-timing-function);
        }
        .mouse-stalker-root[data-st-icon-animation-name-target=\\"toTop\\"].is-hover .mouse-stalker-icon-target {
          animation: toTop 0.3s var(--st-transition-timing-function);
        }
        .mouse-stalker-root[data-st-icon-animation-name-target=\\"toBottom\\"].is-hover .mouse-stalker-icon-target {
          animation: toBottom 0.3s var(--st-transition-timing-function);
        }
        .mouse-stalker-root[data-st-icon-animation-name-target=\\"toRight\\"].is-hover .mouse-stalker-icon-target {
          animation: toRight 0.3s var(--st-transition-timing-function);
        }
        .mouse-stalker-root[data-st-icon-animation-name-target=\\"toLeft\\"].is-hover .mouse-stalker-icon-target {
          animation: toLeft 0.3s var(--st-transition-timing-function);
        }
      }
      
      @keyframes show {
        0% {
          opacity: 0;
          transform: scale(2);
        }
        100% {
          opacity: 1;
          transform: scale(1);
        }
      }
      @keyframes toBottom {
        0% {
          opacity: 0;
          transform: translate(0, -100%);
        }
        100% {
          opacity: 1;
          transform: translate(0, 0);
        }
      }
      @keyframes toTop {
        0% {
          opacity: 0;
          transform: translate(0, 100%);
        }
        100% {
          opacity: 1;
          transform: translate(0, 0);
        }
      }
      @keyframes toRight {
        0% {
          opacity: 0;
          transform: translate(-100%, 0);
        }
        100% {
          opacity: 1;
          transform: translate(0, 0);
        }
      }
      @keyframes toLeft {
        0% {
          opacity: 0;
          transform: translate(100%, 0);
        }
        100% {
          opacity: 1;
          transform: translate(0, 0);
        }
      }
    </style>
  "
`;
